File: apps/mobile-app/ios/VaultStoreKit/Services/VaultSyncEngine.swift
import Foundation
import RustCoreFramework
import VaultModels
import VaultUtils

/// Drives one operation of the Rust vault sync engine against this vault store.
///
/// The engine owns the sync algorithm and emits commands (HTTP, state, SQLite, vault storage); this class
/// carries each one out and feeds the response back until the engine reports `done`. See the `vault_sync`
/// module in `core/rust` for the command and response contract.
public final class VaultSyncEngine {
    /// Prefix of the UserDefaults keys the engine's persisted state lives under.
    private static let statePrefix = "aliasvault_sync_state:"

    /// Engine state keys mirrored into the native account-key chain, so the native password unlock can unwrap it.
    private static let encryptedAccountKeyStateKey = "encryptedAccountKey"
    private static let encryptedVekStateKey = "encryptedVek"
    private static let encryptedAccountPrivateKeyStateKey = "encryptedAccountPrivateKey"

    /// Engine state key that lives in the native store instead (the login writes it there), routed on read and write.
    private static let derivationParamsStateKey = "encryptionKeyDerivationParams"

    /// Engine state key holding the id of the user's personal manifest, which every new row is stamped with.
    public static let personalManifestIdStateKey = "vaultPersonalManifestId"

    private let vaultStore: VaultStore
    private let webApiService: WebApiService
    /// The engine's staging database, held in the Rust core's memory.
    private var staging: SqliteMemoryDatabase?
    private var runLog = VaultSyncRunLog(operation: "")

    public init(vaultStore: VaultStore, webApiService: WebApiService) {
        self.vaultStore = vaultStore
        self.webApiService = webApiService
    }

    deinit {
        closeStaging()
    }

    /// Run one engine operation (`fullSync`, `statusCheck`, `migrationStatus`, `migrateManifest`, `resolveVaultKey`) and
    /// return its result. `encryptionKey` overrides the store's key for the one operation that runs before it is known.
    public func run(operation: String, forcePull: Bool = false, encryptionKey: String? = nil) async throws -> [String: Any] {
        let log = VaultSyncRunLog(operation: operation)
        runLog = log
        var finalResult: [String: Any]?
        let session = try VaultSyncSession(requestJson: try buildRequest(operation: operation, forcePull: forcePull, encryptionKey: encryptionKey))
        defer {
            closeStaging()
            log.finish(result: finalResult, userDefaults: vaultStore.userDefaults)
        }

        while true {
            let commandJson = try log.engine { try session.nextCommand() }
            let command = try log.json { try Self.parseJson(commandJson) }
            let kind = command["kind"] as? String ?? ""
            if kind == "done" {
                let result = command["result"] as? [String: Any] ?? [:]
                finalResult = result
                return result
            }
            let commandStartedAt = Date()
            let response = await handle(kind: kind, command: command)
            log.recordCommand(kind, command: command, response: response, since: commandStartedAt)
            let responseJson = try log.json { try Self.serializeJson(response) }
            try log.engine { try session.resume(responseJson: responseJson) }
        }
    }

    // MARK: - Request

    private func buildRequest(operation: String, forcePull: Bool, encryptionKey: String?) throws -> String {
        let metadata = vaultStore.getVaultMetadataObject()
        let syncState = vaultStore.getSyncState()
        var request: [String: Any] = [
            "operation": operation,
            "username": vaultStore.getUsername() ?? "",
            "isDirty": syncState.isDirty,
            "mutationSequence": syncState.mutationSequence,
            "dirtyScopes": syncState.isDirty ? ["Main"] : [],
            "privateEmailDomains": metadata?.privateEmailDomains ?? [],
            "forcePull": forcePull,
            "minServerVersion": AppInfo.minServerVersion,
            "isOfflineMode": vaultStore.getOfflineMode(),
            "unnamedSharedVaultName": NSLocalizedString("unnamed_shared_vault", value: "Shared vault", comment: "Name of a shared vault without a name")
        ]
        if let key = encryptionKey ?? (try? vaultStore.getEncryptionKeyBase64()) {
            request["encryptionKey"] = key
        }
        if let publicKey = state(forKey: "accountPublicKey") as? String {
            request["accountPublicKey"] = publicKey
        }
        if let privateKey = vaultStore.accountPrivateKey {
            request["accountPrivateKey"] = privateKey
        }
        return try Self.serializeJson(request)
    }

    // MARK: - Commands

    private func handle(kind: String, command: [String: Any]) async -> [String: Any] {
        do {
            switch kind {
            case "http":
                return await http(command)
            case "stateGet":
                return ["value": state(forKey: command["key"] as? String ?? "") ?? NSNull()]
            case "stateSet":
                setState(command["value"] ?? NSNull(), forKey: command["key"] as? String ?? "")
                return [:]
            case "stateRemove":
                setState(nil, forKey: command["key"] as? String ?? "")
                return [:]
            case "dbOpen":
                try openStaging(bytesBase64: command["bytes"] as? String)
                return [:]
            case "dbQuery":
                let db = try database(named: command["db"] as? String ?? "")
                let rowsJson = try db.query(sql: command["sql"] as? String ?? "", paramsJson: try Self.serializeJson(command["params"] ?? []))
                return ["rows": try Self.parseJsonArray(rowsJson)]
            case "dbExec":
                let db = try database(named: command["db"] as? String ?? "")
                try db.exec(statementsJson: try Self.serializeJson(command["statements"] ?? []))
                return [:]
            case "dbExport":
                let bytes: Data
                switch command["db"] as? String ?? "" {
                case "local":
                    bytes = try vaultStore.exportDatabase()
                case "staging":
                    bytes = try database(named: "staging").export()
                case let name:
                    throw AppError.unknownError(message: "Unknown database \(name)")
                }
                return ["bytes": bytes.base64EncodedString()]
            case "vaultStore":
                return try storeVault(command)
            case "vaultLoad":
                return ["encryptedBlob": vaultStore.getEncryptedDatabase() ?? NSNull()]
            case "markClean":
                let cleared = vaultStore.markVaultClean(mutationSeqAtStart: command["mutationSeqAtStart"] as? Int ?? 0, newServerRevision: vaultStore.getCurrentVaultRevisionNumber())
                return ["cleared": cleared]
            case "log":
                runLog.engineLine(level: command["level"] as? String ?? "log", message: command["message"] as? String ?? "")
                return [:]
            default:
                return ["error": "Unknown engine command \(kind)"]
            }
        } catch let error as AppError {
            return ["error": error.message]
        } catch {
            return ["error": "\(error)"]
        }
    }

    private func http(_ command: [String: Any]) async -> [String: Any] {
        let method = command["method"] as? String ?? "GET"
        let path = command["path"] as? String ?? ""
        let body = command["body"] as? String
        let auth = command["auth"] as? Bool ?? true
        var headers: [String: String] = ["Accept": "application/json"]
        if body != nil {
            headers["Content-Type"] = "application/json"
        }
        do {
            let response = try await webApiService.executeRequest(method: method, endpoint: path, body: body, headers: headers, requiresAuth: auth)
            return ["status": response.statusCode, "body": response.body]
        } catch {
            let timedOut = (error as? URLError)?.code == .timedOut
            return ["status": 0, "transportError": error.localizedDescription, "timedOut": timedOut]
        }
    }

    // MARK: - State

    private func state(forKey key: String) -> Any? {
        if key == Self.derivationParamsStateKey {
            guard let json = vaultStore.getUnlockKeyDerivationParams(), let data = json.data(using: .utf8) else { return nil }
            return try? JSONSerialization.jsonObject(with: data, options: [.fragmentsAllowed])
        }
        return Self.persistedState(forKey: key, in: vaultStore.userDefaults)
    }

    /// One persisted engine value, decoded from the JSON text it is stored as.
    public static func persistedState(forKey key: String, in userDefaults: UserDefaults) -> Any? {
        guard let json = userDefaults.string(forKey: statePrefix + key), let data = json.data(using: .utf8) else {
            return nil
        }
        return try? JSONSerialization.jsonObject(with: data, options: [.fragmentsAllowed])
    }

    private func setState(_ value: Any?, forKey key: String) {
        if key == Self.derivationParamsStateKey {
            if let value = value, !(value is NSNull), let data = try? JSONSerialization.data(withJSONObject: value), let json = String(data: data, encoding: .utf8) {
                try? vaultStore.storeUnlockKeyDerivationParams(json)
            }
            return
        }
        if let value = value, !(value is NSNull), let data = try? JSONSerialization.data(withJSONObject: value, options: [.fragmentsAllowed]) {
            vaultStore.userDefaults.set(String(data: data, encoding: .utf8), forKey: Self.statePrefix + key)
        } else {
            vaultStore.userDefaults.removeObject(forKey: Self.statePrefix + key)
        }
        if [Self.encryptedAccountKeyStateKey, Self.encryptedVekStateKey, Self.encryptedAccountPrivateKeyStateKey].contains(key) {
            mirrorAccountKeyChain()
        }
    }

    /// Forget every persisted engine value: revisions, fingerprints, blob cache and the cached key chain.
    public static func clearPersistedState(in userDefaults: UserDefaults) {
        for key in userDefaults.dictionaryRepresentation().keys where key.hasPrefix(statePrefix) {
            userDefaults.removeObject(forKey: key)
        }
    }

    /// Keep the native account-key chain in step with the engine's cached key blobs.
    private func mirrorAccountKeyChain() {
        guard let encryptedAccountKey = state(forKey: Self.encryptedAccountKeyStateKey) as? String, let encryptedVek = state(forKey: Self.encryptedVekStateKey) as? String else {
            vaultStore.storeAccountKeyChain(nil)
            return
        }
        var chain: [String: String] = ["encryptedAccountKey": encryptedAccountKey, "encryptedVek": encryptedVek]
        if let encryptedAccountPrivateKey = state(forKey: Self.encryptedAccountPrivateKeyStateKey) as? String {
            chain["encryptedAccountPrivateKey"] = encryptedAccountPrivateKey
        }
        vaultStore.storeAccountKeyChain(try? Self.serializeJson(chain))
    }

    // MARK: - Vault storage

    private func storeVault(_ command: [String: Any]) throws -> [String: Any] {
        guard let encryptedBlob = command["encryptedBlob"] as? String else {
            throw AppError.vaultStoreFailed(message: "vaultStore command without a blob")
        }
        let result = try vaultStore.storeEncryptedVaultWithSyncState(
            encryptedVault: encryptedBlob,
            markDirty: command["markDirty"] as? Bool ?? false,
            serverRevision: command["revision"] as? Int,
            expectedMutationSeq: command["expectedMutationSeq"] as? Int
        )
        if result.success && vaultStore.isVaultUnlocked {
            // The contract: after a store, the live database is the vault just stored.
            let reloadStartedAt = Date()
            try vaultStore.unlockVault()
            runLog.note("Live vault reloaded (decrypt and open) in \(VaultSyncRunLog.elapsedMs(since: reloadStartedAt))ms")
        }
        return ["success": result.success, "mutationSequence": result.mutationSequence]
    }

    // MARK: - SQLite

    private func database(named name: String) throws -> SqliteMemoryDatabase {
        switch name {
        case "local":
            guard let connection = vaultStore.dbConnection else {
                throw AppError.unknownError(message: "The vault is not unlocked")
            }
            return connection
        case "staging":
            guard let staging = staging else {
                throw AppError.unknownError(message: "The staging database is not open")
            }
            return staging
        default:
            throw AppError.unknownError(message: "Unknown database \(name)")
        }
    }

    /// Open the staging database in memory: from SQLite bytes, or fresh with the current client schema.
    private func openStaging(bytesBase64: String?) throws {
        closeStaging()
        let opened: SqliteMemoryDatabase
        if let bytesBase64 = bytesBase64, let bytes = Data(base64Encoded: bytesBase64) {
            opened = try SqliteMemoryDatabase.fromBytes(bytes: bytes)
        } else {
            opened = try SqliteMemoryDatabase.withSchema(schemaSql: VaultSql.completeSchema)
        }
        // The schema script ends by turning foreign keys on; the engine inserts rows in codec order, not FK order.
        try opened.executeBatch(sql: "PRAGMA foreign_keys = OFF")
        staging = opened
    }

    private func closeStaging() {
        staging = nil
    }

    // MARK: - JSON

    private static func parseJson(_ json: String) throws -> [String: Any] {
        guard let data = json.data(using: .utf8), let object = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            throw AppError.parseError(message: "Engine command is not a JSON object")
        }
        return object
    }

    private static func parseJsonArray(_ json: String) throws -> [Any] {
        guard let data = json.data(using: .utf8), let array = try JSONSerialization.jsonObject(with: data) as? [Any] else {
            throw AppError.parseError(message: "Engine rows are not a JSON array")
        }
        return array
    }

    private static func serializeJson(_ object: Any) throws -> String {
        let data = try JSONSerialization.data(withJSONObject: object, options: [.fragmentsAllowed])
        guard let json = String(data: data, encoding: .utf8) else {
            throw AppError.parseError(message: "Failed to encode JSON")
        }
        return json
    }
}

File: apps/mobile-app/android/app/src/main/java/net/aliasvault/app/vaultstore/VaultSyncEngine.kt
package net.aliasvault.app.vaultstore

import android.database.Cursor
import android.util.Base64
import android.util.Log
import io.requery.android.database.sqlite.SQLiteDatabase
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import net.aliasvault.app.rustcore.JnaInitializer
import net.aliasvault.app.utils.AppInfo
import net.aliasvault.app.vaultstore.models.VaultSql
import net.aliasvault.app.vaultstore.storageprovider.StorageProvider
import net.aliasvault.app.webapi.WebApiService
import org.json.JSONArray
import org.json.JSONObject
import uniffi.aliasvault_core.VaultSyncSession
import java.io.File
import java.net.SocketTimeoutException

/**
 * Drives one operation of the Rust vault sync engine against this vault store.
 *
 * The engine owns the sync algorithm and emits commands (HTTP, state, SQLite, vault storage); this class carries
 * each one out and feeds the response back until the engine reports `done`. See the `vault_sync` module in
 * `core/rust` for the command and response contract.
 */
class VaultSyncEngine(
    private val vaultStore: VaultStore,
    private val storageProvider: StorageProvider,
    private val webApiService: WebApiService,
) {
    companion object {
        private const val TAG = "VaultSyncEngine"

        /** Engine state keys mirrored into the native account-key chain, so the native password unlock can unwrap it. */
        private const val ENCRYPTED_ACCOUNT_KEY_STATE_KEY = "encryptedAccountKey"
        private const val ENCRYPTED_VEK_STATE_KEY = "encryptedVek"

        /** Engine state key that lives in the native store instead (the login writes it there), routed on read and write. */
        private const val DERIVATION_PARAMS_STATE_KEY = "encryptionKeyDerivationParams"
    }

    private var staging: SQLiteDatabase? = null
    private var stagingFile: File? = null
    private var runLog = VaultSyncRunLog("")

    /**
     * Run one engine operation (`fullSync`, `statusCheck`, `migrationStatus`, `migrateManifest`, `resolveVaultKey`) and
     * return its result. [encryptionKey] overrides the store's key for the one operation that runs before it is known.
     */
    suspend fun run(operation: String, forcePull: Boolean = false, encryptionKey: String? = null): JSONObject {
        JnaInitializer.ensureInitialized()
        val log = VaultSyncRunLog(operation)
        runLog = log
        var success: Boolean? = null
        val session = VaultSyncSession(buildRequest(operation, forcePull, encryptionKey))
        try {
            while (true) {
                val commandJson = log.engine { session.nextCommand() }
                val command = log.json { JSONObject(commandJson) }
                val kind = command.optString("kind")
                if (kind == "done") {
                    val result = command.optJSONObject("result") ?: JSONObject()
                    success = if (result.has("success")) result.optBoolean("success") else null
                    return result
                }
                val startNanos = System.nanoTime()
                val response = handle(kind, command)
                log.recordCommand(kind, command, response, System.nanoTime() - startNanos)
                val responseJson = log.json { response.toString() }
                log.engine { session.resume(responseJson) }
            }
        } finally {
            closeStaging()
            session.destroy()
            log.finish(success, storageProvider)
        }
    }

    // region Request

    private fun buildRequest(operation: String, forcePull: Boolean, encryptionKey: String?): String {
        val metadata = vaultStore.metadata.getVaultMetadataObject()
        val syncState = vaultStore.getSyncState()
        val request = JSONObject().apply {
            put("operation", operation)
            put("username", vaultStore.getUsername() ?: "")
            put("isDirty", syncState.isDirty)
            put("mutationSequence", syncState.mutationSequence)
            put("dirtyScopes", JSONArray(if (syncState.isDirty) listOf("Main") else emptyList<String>()))
            put("privateEmailDomains", JSONArray(metadata?.privateEmailDomains ?: emptyList<String>()))
            put("forcePull", forcePull)
            put("minServerVersion", AppInfo.MIN_SERVER_VERSION)
            put("isOfflineMode", vaultStore.getOfflineMode())
            put("unnamedSharedVaultName", "Shared vault")
        }
        (encryptionKey ?: vaultStore.getEncryptionKeyBase64())?.let { request.put("encryptionKey", it) }
        (state("accountPublicKey") as? String)?.let { request.put("accountPublicKey", it) }
        vaultStore.accountPrivateKey?.let { request.put("accountPrivateKey", it) }
        return request.toString()
    }

    // endregion

    // region Commands

    @Suppress("TooGenericExceptionCaught")
    private suspend fun handle(kind: String, command: JSONObject): JSONObject {
        return try {
            when (kind) {
                "http" -> http(command)
                "stateGet" -> JSONObject().put("value", state(command.optString("key")) ?: JSONObject.NULL)
                "stateSet" -> {
                    setState(command.optString("key"), command.opt("value"))
                    JSONObject()
                }
                "stateRemove" -> {
                    setState(command.optString("key"), null)
                    JSONObject()
                }
                "dbOpen" -> {
                    openStaging(command.optString("bytes", null).takeIf { command.has("bytes") && !command.isNull("bytes") })
                    JSONObject()
                }
                "dbQuery" -> {
                    val rows = query(database(command.optString("db")), command.optString("sql"), command.optJSONArray("params") ?: JSONArray())
                    JSONObject().put("rows", rows)
                }
                "dbExec" -> {
                    exec(database(command.optString("db")), command.optJSONArray("statements") ?: JSONArray())
                    JSONObject()
                }
                "dbExport" -> JSONObject().put("bytes", Base64.encodeToString(export(database(command.optString("db"))), Base64.NO_WRAP))
                "vaultStore" -> storeVault(command)
                "vaultLoad" -> JSONObject().put("encryptedBlob", if (vaultStore.hasEncryptedDatabase()) vaultStore.getEncryptedDatabase() else JSONObject.NULL)
                "markClean" -> {
                    val cleared = vaultStore.markVaultClean(command.optInt("mutationSeqAtStart"), vaultStore.metadata.getVaultRevisionNumber())
                    JSONObject().put("cleared", cleared)
                }
                "log" -> {
                    runLog.engineLine(command.optString("level", "log"), command.optString("message"))
                    JSONObject()
                }
                else -> JSONObject().put("error", "Unknown engine command $kind")
            }
        } catch (e: Exception) {
            Log.w(TAG, "Engine command $kind failed", e)
            JSONObject().put("error", e.message ?: e.toString())
        }
    }

    @Suppress("TooGenericExceptionCaught")
    private suspend fun http(command: JSONObject): JSONObject = withContext(Dispatchers.IO) {
        val body = if (command.has("body") && !command.isNull("body")) command.getString("body") else null
        val headers = mutableMapOf("Accept" to "application/json")
        if (body != null) {
            headers["Content-Type"] = "application/json"
        }
        try {
            val response = webApiService.executeRequest(
                method = command.optString("method", "GET"),
                endpoint = command.optString("path"),
                body = body,
                headers = headers,
                requiresAuth = command.optBoolean("auth", true),
            )
            JSONObject().put("status", response.statusCode).put("body", response.body)
        } catch (e: SocketTimeoutException) {
            JSONObject().put("status", 0).put("transportError", e.toString()).put("timedOut", true)
        } catch (e: Exception) {
            JSONObject().put("status", 0).put("transportError", e.toString()).put("timedOut", false)
        }
    }

    // endregion

    // region State

    /** Engine state is stored as the JSON text of `{"v": value}`, so any JSON value round-trips. */
    private fun state(key: String): Any? {
        if (key == DERIVATION_PARAMS_STATE_KEY) {
            return storageProvider.getKeyDerivationParams().takeIf { it.isNotBlank() }?.let { JSONObject(it) }
        }
        val json = storageProvider.getSyncEngineState(key) ?: return null
        return JSONObject(json).opt("v")?.takeIf { it != JSONObject.NULL }
    }

    private fun setState(key: String, value: Any?) {
        if (key == DERIVATION_PARAMS_STATE_KEY) {
            (value as? JSONObject)?.let { storageProvider.setKeyDerivationParams(it.toString()) }
            return
        }
        if (value == null || value == JSONObject.NULL) {
            storageProvider.setSyncEngineState(key, null)
        } else {
            storageProvider.setSyncEngineState(key, JSONObject().put("v", value).toString())
        }
        if (key == ENCRYPTED_ACCOUNT_KEY_STATE_KEY || key == ENCRYPTED_VEK_STATE_KEY) {
            mirrorAccountKeyChain()
        }
    }

    /**
     * Keep the native account-key chain in step with the engine's cached key blobs.
     */
    private fun mirrorAccountKeyChain() {
        val encryptedAccountKey = state(ENCRYPTED_ACCOUNT_KEY_STATE_KEY) as? String
        val encryptedVek = state(ENCRYPTED_VEK_STATE_KEY) as? String
        if (encryptedAccountKey == null || encryptedVek == null) {
            vaultStore.storeAccountKeyChain(null)
            return
        }
        val chain = JSONObject().put("encryptedAccountKey", encryptedAccountKey).put("encryptedVek", encryptedVek)
        vaultStore.storeAccountKeyChain(chain.toString())
    }

    // endregion

    // region Vault storage

    private fun storeVault(command: JSONObject): JSONObject {
        val encryptedBlob = command.getString("encryptedBlob")
        if (command.has("encryptionKey") && !command.isNull("encryptionKey")) {
            // The blob is encrypted under a key this session did not start with (KEK to VEK migration): adopt it first.
            vaultStore.adoptEncryptionKey(command.getString("encryptionKey"))
        }
        val result = vaultStore.storeEncryptedVaultWithSyncState(
            encryptedVault = encryptedBlob,
            markDirty = command.optBoolean("markDirty", false),
            serverRevision = optionalInt(command, "revision"),
            expectedMutationSeq = optionalInt(command, "expectedMutationSeq"),
        )
        if (result.success && vaultStore.isVaultUnlocked()) {
            // The contract: after a store, the live database is the vault just stored.
            val reloadStartNanos = System.nanoTime()
            vaultStore.unlockVault()
            runLog.note("Live vault reloaded (decrypt and open) in ${VaultSyncRunLog.elapsedMsSince(reloadStartNanos)}ms")
        }
        return JSONObject().put("success", result.success).put("mutationSequence", result.mutationSequence)
    }

    private fun optionalInt(command: JSONObject, key: String): Int? {
        return if (command.has(key) && !command.isNull(key)) command.getInt(key) else null
    }

    // endregion

    // region SQLite

    private fun database(name: String): SQLiteDatabase = when (name) {
        "local" -> vaultStore.database.dbConnection ?: error("The vault is not unlocked")
        "staging" -> staging ?: error("The staging database is not open")
        else -> error("Unknown database $name")
    }

    /**
     * Open the staging database: from SQLite bytes (as a temporary file), or fresh in memory with the current schema.
     */
    private fun openStaging(bytesBase64: String?) {
        closeStaging()
        if (bytesBase64 != null) {
            val file = File(storageProvider.getRandomTempFilePath())
            file.writeBytes(Base64.decode(bytesBase64, Base64.NO_WRAP))
            stagingFile = file
            staging = SQLiteDatabase.openDatabase(file.path, null, SQLiteDatabase.OPEN_READWRITE).also {
                it.rawQuery("PRAGMA foreign_keys = OFF", null)?.close()
            }
        } else {
            staging = SQLiteDatabase.create(null).also {
                executeScript(it, VaultSql.completeSchema)
                // The schema script ends by turning foreign keys on; the engine inserts rows in codec order, not FK order.
                it.rawQuery("PRAGMA foreign_keys = OFF", null)?.close()
            }
        }
    }

    private fun closeStaging() {
        staging?.close()
        staging = null
        stagingFile?.delete()
        stagingFile = null
    }

    /**
     * Run a multi-statement SQL script. Statements split on semicolons, except inside a trigger body.
     */
    private fun executeScript(db: SQLiteDatabase, script: String) {
        for (statement in SqlScript.splitStatements(script)) {
            val upper = statement.uppercase()
            when {
                upper.startsWith("PRAGMA") -> db.rawQuery(statement, null)?.close()
                upper.startsWith("BEGIN") || upper.startsWith("COMMIT") || upper.startsWith("ROLLBACK") -> db.execSQL(statement)
                else -> db.compileStatement(statement).use { it.execute() }
            }
        }
    }

    private fun export(db: SQLiteDatabase): ByteArray {
        val file = File(storageProvider.getRandomTempFilePath())
        file.delete()
        try {
            val quoted = file.absolutePath.replace("'", "''")
            db.compileStatement("VACUUM INTO '$quoted'").use { it.execute() }
            return file.readBytes()
        } finally {
            file.delete()
        }
    }

    private fun bindValue(value: Any?): Any? = when (value) {
        null, JSONObject.NULL -> null
        is Boolean -> if (value) 1L else 0L
        is Int -> value.toLong()
        is Long -> value
        is Double -> value
        is Float -> value.toDouble()
        is String -> value
        is JSONObject -> if (value.has("__b64")) Base64.decode(value.getString("__b64"), Base64.NO_WRAP) else value.toString()
        else -> value.toString()
    }

    private fun query(db: SQLiteDatabase, sql: String, params: JSONArray): JSONArray {
        val args = Array<Any?>(params.length()) { bindValue(params.opt(it)) }
        val rows = JSONArray()
        db.rawQuery(sql, args).use { cursor ->
            while (cursor.moveToNext()) {
                val row = JSONObject()
                for (index in 0 until cursor.columnCount) {
                    val name = cursor.getColumnName(index)
                    when (cursor.getType(index)) {
                        Cursor.FIELD_TYPE_INTEGER -> row.put(name, cursor.getLong(index))
                        Cursor.FIELD_TYPE_FLOAT -> row.put(name, cursor.getDouble(index))
                        Cursor.FIELD_TYPE_STRING -> row.put(name, cursor.getString(index))
                        Cursor.FIELD_TYPE_BLOB -> row.put(name, JSONObject().put("__b64", Base64.encodeToString(cursor.getBlob(index), Base64.NO_WRAP)))
                        else -> row.put(name, JSONObject.NULL)
                    }
                }
                rows.put(row)
            }
        }
        return rows
    }

    private fun exec(db: SQLiteDatabase, statements: JSONArray) {
        db.beginTransaction()
        try {
            for (index in 0 until statements.length()) {
                val entry = statements.getJSONObject(index)
                db.compileStatement(entry.getString("sql")).use { statement ->
                    bindParams(statement, entry.optJSONArray("params") ?: JSONArray())
                    statement.execute()
                }
            }
            db.setTransactionSuccessful()
        } finally {
            db.endTransaction()
        }
    }

    private fun bindParams(statement: io.requery.android.database.sqlite.SQLiteStatement, params: JSONArray) {
        for (position in 0 until params.length()) {
            when (val value = bindValue(params.opt(position))) {
                null -> statement.bindNull(position + 1)
                is Long -> statement.bindLong(position + 1, value)
                is Double -> statement.bindDouble(position + 1, value)
                is ByteArray -> statement.bindBlob(position + 1, value)
                else -> statement.bindString(position + 1, value.toString())
            }
        }
    }

    // endregion
}
